walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,75 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Logger;
 
+/* FindMeetingQuery determines if people can have meetings based on each persons avaliable times */
 public final class FindMeetingQuery {
+
+  private static Logger log = Logger.getLogger(FindMeetingQuery.class.getName());
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<TimeRange> originalAvailableTimes = Arrays.asList(TimeRange.WHOLE_DAY);
+    HashSet<String> attendees = new HashSet(request.getAttendees());
+    HashSet<String> eventAttendees = new HashSet<String>();
+    
+    // If meeting is longer than the whole day, return empty list 
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees, return TimeRange of whole day 
+    if (attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);  
+    }
+
+    for (Event event : events) {
+      
+      Collection<TimeRange> availableTimesAroundEvent = new ArrayList<TimeRange>();
+      eventAttendees.addAll(event.getAttendees());  
+      
+      for (TimeRange availableTime : originalAvailableTimes) {
+
+        TimeRange eventTime = event.getWhen();
+
+        if (!eventTime.overlaps(availableTime)) {
+          availableTimesAroundEvent.add(availableTime);
+          continue;
+        }
+
+        // Creates TimeRange object of the available meeting time range before the event starts 
+        if (eventTime.start() > availableTime.start()) {
+          TimeRange availableTimeBeforeEvent = 
+            TimeRange.fromStartEnd(availableTime.start(), eventTime.start(), false);
+          if (availableTimeBeforeEvent.duration() >= request.getDuration()) {
+            availableTimesAroundEvent.add(availableTimeBeforeEvent);
+          }
+        }
+
+        // Creates TimeRange object of the available meeting time range after the event ends 
+        if (availableTime.end() > eventTime.end()) {
+          TimeRange availableTimeAfterEvent = 
+            TimeRange.fromStartEnd(eventTime.end(), availableTime.end(), false);
+          if (availableTimeAfterEvent.duration() >= request.getDuration()) {
+            availableTimesAroundEvent.add(availableTimeAfterEvent);
+          } 
+        }
+      }
+      originalAvailableTimes = availableTimesAroundEvent;
+    }
+    boolean checkAttendees = eventAttendees.equals(attendees);
+    if(checkAttendees) {
```

Comment:
nit: need space after if

i would probably do something like this:

```
// If the attendees of the events match the attendees in the request, then we can return the available times. 
// If there's a mismatch, just return the default. 
if (eventAttendees.equals(attendees)) {
  return originalAvailableTimes;
}

return Arrays.asList(TimeRange.WHOLE_DAY);
```

Reply:
also does this check make the test pass? i think ideally you should only be splitting available times when an existing event contains a required attendee.